usvm-python/usvm-python-main/src/main/kotlin/org/usvm/machine/PyState.kt

```diff
@@ -0,0 +1,151 @@
+    var delayedForks: PersistentList<DelayedFork> = persistentListOf(),
+    private val mocks: MutableMap<MockHeader, UMockSymbol<UAddressSort>> = mutableMapOf(),
+    val mockedObjects: MutableSet<UninterpretedSymbolicPythonObject> = mutableSetOf(),
+    var uniqueInstructions: PersistentSet<PyInstruction> = persistentSetOf(),
+) : UState<PythonType, PyCallable, PyInstruction, PyContext, PyTarget, PyState>(
+    ctx,
+    callStack,
+    pathConstraints,
+    memory,
+    listOf(uModel),
+    pathLocation,
+    forkPoints,
+    targets,
+) {
+    override fun clone(newConstraints: UPathConstraints<PythonType>?): PyState {
+        require(newConstraints is PyPathConstraints?)
+        val newPathConstraints = newConstraints ?: pathConstraints.clone()
+        val newMemory = memory.clone(newPathConstraints.typeConstraints)
+        return PyState(
+            ctx,
+            pythonCallable,
+            inputSymbols,
+            newPathConstraints,
+            newMemory,
+            models.first(),
+            typeSystem,
+            preAllocatedObjects.clone(),
+            possibleTypesForNull,
+            callStack,
+            pathNode,
+            forkPoints,
+            concolicQueries,
+            delayedForks,
+            mocks.toMutableMap(), // copy
+            mockedObjects.toMutableSet(), // copy
+            uniqueInstructions
+        )
+    }
+
+    override val entrypoint = pythonCallable
+    override val isExceptional: Boolean = false // TODO
+    val meta = PythonExecutionStateMeta()
+    val pyModel: PyModel
+        get() = models.first() as? PyModel ?: error("Model PyState must be PyModel")
+    fun buildPathAsList(): List<SymbolicHandlerEvent<Any>> = concolicQueries
+
+    fun mock(what: MockHeader): MockResult {
+        val cached = mocks[what]
+        if (cached != null) {
+            return MockResult(UninterpretedSymbolicPythonObject(cached, typeSystem), false, cached)
+        }
+        val result = memory.mocker.call(what.method, what.args.map { it.address }.asSequence(), ctx.addressSort)
+        mocks[what] = result
+        what.methodOwner?.let { mockedObjects.add(it) }
+        return MockResult(UninterpretedSymbolicPythonObject(result, typeSystem), true, result)
+    }
+
+    fun getMocksForSymbol(
+        symbol: UninterpretedSymbolicPythonObject,
+    ): List<Pair<MockHeader, UninterpretedSymbolicPythonObject>> =
+        mocks.mapNotNull { (mockHeader, mockResult) ->
+            if (mockHeader.methodOwner == symbol) {
+                mockHeader to UninterpretedSymbolicPythonObject(mockResult, typeSystem)
+            } else {
+                null
+            }
+        }
+
+    fun isTerminated(): Boolean {
+        return meta.modelDied || meta.wasInterrupted || meta.wasExecuted && meta.objectsWithoutConcreteTypes == null
+    }
+
+    fun isInterestingForPathSelector(): Boolean {
+        return !isTerminated() || delayedForks.isNotEmpty()
+    }
+}
+
+class DelayedFork(
```

Comment:
`DelayedForkByInstruction.kt` is something completely different. I should probably rename it somehow to avoid confusion.

`DelayedForkByInstruction` is one specific strategy for `PyVirtualPathSelector`. Unlike the very basic implementation, it groups delayed forks by instructions.